change identity comparisons to use isequal

vtjnash opened the pull request:
this makes it now possible to make `==` with non-comparable items an error (ref #15983), since these notions are no longer coupled, but it _does not_ make that change now

~~it does change `in` to use `isequal` (#9381), but that's a small part of the change, so I might move that to another branch while that issue is still under discussion~~ edit: now extracted from this PR to https://github.com/JuliaLang/julia/compare/jn/notequal...jn/in-isequal

vtjnash committed:
change identity comparisons to use isequal

it would now be possible to make == with non-comparable items an error
since these notions are no longer coupled

fix #9381
ref #15983

JeffBezanson commented:
Yes, the behavior change to `in` should definitely be separated; the rest is uncontroversial cleanup to use `===` and `isequal` appropriately.

nalimilan commented on base/array.jl:
```diff
@@ -764,25 +764,27 @@ function findprev(A, start::Integer)
 function findprev(testf::Function, A, start::Integer)
     for i = start:-1:1
-        testf(A[i]) && return i
+        if testf(A[i])
```
Why make this change here and not above? (Anyway, that's clearly unrelated to this PR...)

tkelman commented on base/statistics.jl:
```diff
@@ -161,16 +161,16 @@ varm{T}(A::AbstractArray{T}, m::AbstractArray, region; corrected::Bool=true) =
     mean === nothing ? varm(A, Base.mean(A, region), region; corrected=corrected) :
-    isa(mean, AbstractArray) ? varm(A, mean::AbstractArray, region; corrected=corrected) :
+    isa(mean, AbstractArray) ? varm(A, mean, region; corrected=corrected) :
+    isequal(mean, 0) ? varzm(A, region; corrected=corrected) :
```
would you still want to call varzm on negative 0?

vtjnash replied:
the branching here is odd anyways, and seems to be a relic of an older version of the PR that improved these (https://github.com/JuliaLang/julia/pull/6273#issuecomment-38999552)

tkelman replied:
that isn't an answer to the question

tkelman commented on base/array.jl:
```diff
@@ -740,7 +740,7 @@ findfirst(A) = findnext(A, 1)
 function findnext(A, v, start::Integer)
     for i = start:length(A)
-        if A[i] == v
+        if isequal(A[i], v)
```
this should be documented and tested as a behavior change on findnext/findprev with -0.0

StefanKarpinski commented:
> the rest is uncontroversial cleanup to use === and isequal appropriately.

Agreed: this should be in its own PR and merged immediately.

vtjnash committed:
remove varzm functions

the external abi is to call var,
the internal abi doesn' need to branch to alternative functions based on whether mean is given as zero
that simply made the dispatch less straightforward to understand
even though doing an exact comparison to 0 isn't generally reliable

ref #6273, which initially introduced these functions as the public API, before changing them to be the internal implementation

vtjnash commented:
> merge immediately

I'll merge in the morning, if there's no further comments

JeffBezanson commented on base/sharedarray.jl:
```diff
@@ -148,14 +148,17 @@ function SharedArray{T,N}(filename::AbstractString, ::Type{T}, dims::NTuple{N,In
     mode in ("r", "r+", "w+", "a+") || throw(ArgumentError("mode must be readable, but $mode is not"))
-    init==false || mode in ("r+", "w+", "a+") || throw(ArgumentError("cannot initialize unwritable array (mode = $mode)"))
+    if init !== false
+        typeassert(init, Function)
```
This typeassert seems to add a restriction that was not here before.

vtjnash replied:
good call. I added this so that `init` isn't silently ignored later (https://github.com/JuliaLang/julia/pull/16764/files/e6e1312a051d9b5511bf5b4f9ea0ae97528bb5ed#diff-6ba74fe8f18fa73bec8dfbfd1801a05cR209), but that test is wrong too

tkelman replied:
what are you trying to link to?

nalimilan commented on base/array.jl:
```diff
@@ -650,7 +650,7 @@ end
     c = ccall(:memcmp, Int32, (Ptr{UInt8}, Ptr{UInt8}, UInt),
               a, b, min(length(a),length(b)))
-    c < 0 ? -1 : c > 0 ? +1 : cmp(length(a),length(b))
+    return c < 0 ? -1 : c > 0 ? +1 : cmp(length(a),length(b))
```
Not an objection, but if you/we think it's better style to always use `return`, could you add a mention of that in the style guide?

malmaud commented:
This broke https://github.com/JuliaStats/StatsBase.jl/issues/169, which depends on `varzm`.

tkelman commented:
should either revert https://github.com/JuliaLang/julia/pull/16764/commits/e6e1312a051d9b5511bf5b4f9ea0ae97528bb5ed or add a deprecation. DataArrays uses it too.
